File: 백준/Silver/1992. 쿼드트리/쿼드트리.java
import java.io.*;
import java.util.*;

public class Main {
    static char[][] board;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    
        int n = Integer.parseInt(br.readLine());
        board = new char[n][n];
        for(int i=0;i<n;i++){
            board[i] = br.readLine().toCharArray();
        }
        String res = compress(0,0,n);
        System.out.println(res);
        }
        private static String compress(int y, int x, int size){
            char start = board[y][x];
            for(int i=y;i<y+size;i++){
                for(int j=x;j<x+size;j++){
                    if(board[i][j]!=start){
                        int half = size/2;
                        String topLeft = compress(y, x, half);
                        String topRight = compress(y, x+half, half);
                        String bottomLeft = compress(y+half, x, half);
                        String bottomRight = compress(y+half, x+half, half);

                        return "("+topLeft+topRight+bottomLeft+bottomRight+")";
                    }
                }
            }
            return String.valueOf(start);
        }
        }
    

File: 백준/Silver/1992. 쿼드트리/쿼드트리.py
import sys
input = sys.stdin.readline

def compress(y, x, size):
    start = board[y][x]
    for row in range(y, y + size):
        for col in range(x, x + size):
            if board[row][col] != start:
                half = size // 2
                top_left = compress(y, x, half)
                top_right = compress(y, x + half, half)
                bottom_left = compress(y + half, x, half)
                bottom_right = compress(y + half, x + half, half)
                return f"({top_left}{top_right}{bottom_left}{bottom_right})"
    return start

n = int(input())
board = [list(input().strip()) for _ in range(n)]
result = compress(0, 0, n)
print(result)
